ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs

```diff
@@ -512,51 +547,95 @@ addReprocessLoEBlocks
   => Tracer m (TraceAddBlockEvent blk)
   -> ChainSelQueue m blk
   -> m ()
-addReprocessLoEBlocks tracer (ChainSelQueue queue) = do
+addReprocessLoEBlocks tracer (ChainSelQueue {varChainSelReprocessLoEBlocks}) = do
   traceWith tracer $ AddedReprocessLoEBlocksToQueue
-  atomically $ writeTBQueue queue ChainSelReprocessLoEBlocks
+  atomically $ writeTVar varChainSelReprocessLoEBlocks True
 
 -- | Get the oldest message from the 'ChainSelQueue' queue. Can block when the
 -- queue is empty; in that case, reports the starvation (and its end) to the
 -- callback.
 getChainSelMessage
-  :: (IOLike m, HasHeader blk)
+  :: (HasHeader blk, IOLike m)
   => Tracer m (TraceChainSelStarvationEvent blk)
   -> StrictTVar m ChainSelStarvation
   -> ChainSelQueue m blk
   -> m (ChainSelMessage m blk)
-getChainSelMessage starvationTracer starvationVar (ChainSelQueue queue) = do
-  -- NOTE: The test of emptiness and the blocking read are in different STM
-  -- transactions on purpose.
-  whenM (atomically $ isEmptyTBQueue queue) $ do
-    writeTVarIO starvationVar ChainSelStarvationOngoing
-    traceWith starvationTracer . ChainSelStarvationStarted =<< getMonotonicTime
-  message <- atomically $ readTBQueue queue
-  -- If there was a starvation ongoing, we need to report that it is done.
-  whenM ((== ChainSelStarvationOngoing) <$> readTVarIO starvationVar) $
-    case message of
-      ChainSelAddBlock BlockToAdd {blockToAdd} -> do
-        time <- getMonotonicTime
-        traceWith starvationTracer $ ChainSelStarvationEnded time $ blockRealPoint blockToAdd
-        writeTVarIO starvationVar $ ChainSelStarvationEndedAt time
-      ChainSelReprocessLoEBlocks -> pure ()
-  return message
+getChainSelMessage starvationTracer starvationVar queue = go
   where
+    go = do
+      (reprocessLoEBlocks, chainSelQueue) <- atomically readBoth
+      case reprocessLoEBlocks of
+        True -> do
+          writeTVarIO varChainSelReprocessLoEBlocks False
+          return ChainSelReprocessLoEBlocks
+        False ->
+          case Map.minView chainSelQueue of
+            Just (blockToAdd, chainSelQueue') -> do
+              writeTVarIO varChainSelQueue chainSelQueue'
```

Comment:
There is a race condition here when blocks are added to the `varChainSelQueue` after `readBoth` but before `writeTVarIO varChainSelQueue chainSelQueue'`.

Reply:
See https://github.com/intersectMBO/ouroboros-consensus/commit/23542f9d25b10c696bb967050eefed0e4a139abe